src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,570 @@
+    }
+
+    /**
+     * Derive a key, returned as a {@code SecretKey}.
+     * <p>
+     * The {@code deriveKey} method may be called multiple times once a
+     * {@code KDF} object is initialized.
+     * <p>
+     * Delayed provider selection is also supported such that the provider
+     * performing the derive is not selected until the method is called.
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code SecretKey} object
+     * @param kdfParameterSpec
+     *     derivation parameters
+     *
+     * @return a {@code SecretKey} object corresponding to a key built from the
+     *     KDF output and according to the derivation parameters
+     *
+     * @throws InvalidParameterSpecException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived, or specifies
+     *     a type of output that is not a key (e.g. raw data)
+     */
+    public SecretKey deriveKey(String alg, KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+        synchronized (lock) {
+            if (spi != null) {
+                return spi.engineDeriveKey(alg, kdfParameterSpec);
+            }
+            Exception lastException = null;
+            while ((firstService != null) || serviceIterator.hasNext()) {
+                Service s;
+                if (firstService != null) {
+                    s = firstService;
+                    firstService = null;
+                } else {
+                    s = serviceIterator.next();
+                }
+                if (!JceSecurity.canUseProvider(s.getProvider())) {
+                    continue;
+                }
+                try {
+                    KDFSpi spi = (KDFSpi) s.newInstance(algorithmParameterSpec);
+                    SecretKey result = spi.engineDeriveKey(alg,
+                                                           kdfParameterSpec);
+                    provider = s.getProvider();
+                    this.spi = spi;
+                    firstService = null;
+                    serviceIterator = null;
+                    return result;
+                } catch (Exception e) {
+                    if (lastException == null) {
+                        lastException = e;
+                    }
+                }
+            }
+            // no working provider found, fail
+            if (lastException instanceof InvalidParameterSpecException) {
+                throw (InvalidParameterSpecException) lastException;
+            }
+            if (lastException instanceof RuntimeException) {
+                throw (RuntimeException) lastException;
+            }
+        }
+        // should never reach here
+        return null;
+    }
+
+    /**
+     * Obtain raw data from a key derivation function.
+     * <p>
+     * The {@code deriveData} method may be called multiple times once a
+     * {@code KDF} object is initialized.
```

Comment:
What does "once a KDF object is initialized" mean? The getInstance methods return a KDF that is initialized, are you saying it's possible to KDF that is only partly initialized? Is this about lazy binding to the provider?